Mongo source

pauloouriques opened the pull request.

bleonard committed:
Something like this to add more mongo options

pauloouriques committed:
chore: mongo source conversion.

pauloouriques committed:
chore: convert getchangedrows to mongo query.

bleonard commented on plugins/@grouparoo/app-templates/src/source/table/getExamples.ts:
```diff
@@ -24,6 +26,7 @@ export interface GetColumnExamplesMethod {
   (argument: {
     connection: any;
     appOptions: SimpleAppOptions;
+    sourceOptions?: SimpleSourceOptions;
```
ideally this wouldn't be optional, but we can do that later.

bleonard commented on plugins/@grouparoo/app-templates/src/source/table/meta.ts:
```diff
@@ -98,19 +102,19 @@ export const buildConnection: BuildConnectionMethod = ({
 
+  const options = (additionalOptions?.options || []).concat({
+    key: tableNameKey,
+    displayName: "Table",
```
maybe we need `tableOptionDisplayName` here so it can be "Collection"

bleonard commented on plugins/@grouparoo/mongo/__tests__/table-import/get-changed-rows-count.ts:
```diff
@@ -0,0 +1,31 @@
+const { appId, appOptions } = getConfig();
+
+async function callGetChangedRowsCount() {
+  const connection = await connect({ appOptions, app: null, appId: null });
```
need to call `await client.close();`?

bleonard replied:
it should probably be in a `beforeAll` and `afterAll` in all these tests. Something like
`beforeAll`: `connection = await connect({appOptions})`
`afterAll`: `await disconnect({connection})`

bleonard commented on plugins/@grouparoo/mongo/__tests__/table-import/get-changed-rows.ts:
```diff
@@ -0,0 +1,36 @@
+import { FilterOperation } from "@grouparoo/app-templates/dist/source/table";
+
+// these used and set by test
+const { appId, appOptions } = getConfig();
```
`tableName` here?

bleonard commented on plugins/@grouparoo/mongo/__tests__/table-import/get-columns.ts:
```diff
@@ -0,0 +1,25 @@
+describe("mongodb/column/getColumnsTest", () => {
+  test("gets the list of columns", async () => {
+    const columns = await getMongoColumns();
+    expect(columns.length).toBe(5);
```
check what they are and their types.

bleonard commented on plugins/@grouparoo/mongo/__tests__/table-import/get-property-value.ts:
```diff
@@ -0,0 +1,43 @@
+  });
+}
+
+describe("mongodb/column/getColumnsTest", () => {
```
title not the right one.
probably plenty more tests here for all the types, etc.

bleonard commented on plugins/@grouparoo/mongo/src/lib/query-import/executeQuery.ts:
```diff
@@ -0,0 +1,8 @@
+  connection,
+  query,
+}) => {
+  return connection.execute(query);
```
`connect` returns `{ client, db }` so isn't this `connection.db.query()` or something?
Is `client` actually used or does it already have a point to the `client` ? Just `db` might be better.

bleonard commented on plugins/@grouparoo/mongo/src/lib/table-import/connection.ts:
```diff
@@ -0,0 +1,35 @@
+    name: "mongodb-collection-import",
+    description:
+      "Import or update Profiles from a Mongo DB database collection.",
+    tableOptionDescription: "The table to scan",
```
Probably "The collection to scan"

bleonard commented on plugins/@grouparoo/mongo/src/lib/table-import/getColumns.ts:
```diff
@@ -0,0 +1,139 @@
+}) => {
+  const map: ColumnDefinitionMap = {};
+  if (sourceOptions) {
+    const fields = sourceOptions.fields.toString().split(",");
```
let's also `trim`: `sourceOptions.fields.toString().split(",").map(f => f.trim())`

bleonard replied:
And have a test with spaces.

bleonard commented on plugins/@grouparoo/mongo/src/lib/table-import/getColumns.ts:
```diff
@@ -0,0 +1,139 @@
+
+    // other?
+    default:
+      break;
```
what if it's not found. How do other plugins handle this?

bleonard replied:
I notice postgres just says `null` so maybe core handles it. let's check.

bleonard commented on plugins/@grouparoo/mongo/src/lib/table-import/getSampleRows.ts:
```diff
@@ -0,0 +1,28 @@
+}) => {
+  const out: DataResponseRow[] = [];
+  if (sourceOptions) {
+    const fields = sourceOptions.fields.toString().split(",");
```
`trim` here too. maybe `DRY` that up.

bleonard replied:
There were spots before when you were replacing the tbale name and such. is that always the case?

bleonard commented on plugins/@grouparoo/app-templates/src/source/table/pluginMethods.ts:
```diff
@@ -29,6 +34,7 @@ export enum FilterOperation {
   Contain = "contains",
   NotContain = "does not contain",
   In = "in",
+  Exists = "exists",
```
I don't think this is used.

bleonard commented on plugins/@grouparoo/mongo/src/lib/table-import/getColumns.ts:
```diff
@@ -0,0 +1,139 @@
+    ops.push(FilterOperation.LessThanOrEqual);
+  }
+  if (contains) {
+    // ops.push(MongoFilterOperation.Contain);
```
Is this supported by Mongo?

bleonard replied:
If they are, these would be `FilterOperation` not `MongoFilterOperation`

bleonard commented on plugins/@grouparoo/mongo/__tests__/utils/data.ts:
```diff
@@ -0,0 +1,210 @@
+          profile_id: {
+            bsonType: "int",
+          },
+          purchase: {
```
maybe add a nested object and array in here for testing the dot syntax

pauloouriques committed:
chore: mongo source.

 - Add mongo source;
 - Add sourceOptions as optional parameter where needed;
 - Add collection/query import tests.

pauloouriques committed:
chore: last query import tests.

grouparoo-bot committed:
Updating things automatically

pauloouriques committed:
chore: last adjusts and circleci mongo image.

pauloouriques committed:
chore: last adjusts and add circle ci mongo image.

grouparoo-bot committed:
Updating things automatically

pauloouriques committed:
feat: add mongo image.

grouparoo-bot committed:
Updating things automatically

pauloouriques committed:
fix: close mongo connection issue,

pauloouriques committed:
feat: add nested field support and tests.

pauloouriques committed:
chore: bump Mongo plogin vestion up.

pauloouriques committed:
fix: core and app-templates dep versions issue.

pauloouriques committed:
fix: uncomment after data.

pauloouriques committed:
fix: change GROUPAROO_INJECTED_PLUGINS to mongo.

pauloouriques committed:
fix: add timezone to dates since the constructor is using epoch.

pauloouriques committed:
chore: bump mongo plugin version up.

evantahler commented:
To install and run Mongo on OSX (via homebrew)

```
brew tap mongodb/brew
brew install mongodb-community
brew services start mongodb/brew/mongodb-community
```

From https://docs.mongodb.com/manual/tutorial/install-mongodb-on-os-x/

evantahler commented on plugins/@grouparoo/mongo/package.json:
```diff
@@ -0,0 +1,48 @@
+  },
+  "dependencies": {
+    "@grouparoo/app-templates": "0.2.8",
+    "mongodb": "^3.6.4"
```
Please don't use the `^` and choose a specific version

evantahler commented on plugins/@grouparoo/mongo/public/templates/app/{{{id}}}.js.template:
```diff
@@ -0,0 +1,14 @@
+      type: "{{{__pluginName}}}",
+      options: {
+        uri: "mongodb://localhost:27017/mydb", // The MongoDB Connection String.
+        database: "mydb", // The MongoDB database name (normally, the Connection String have the database name too).
```
Does this mean that this is optional because it will be in the uri?  If so, perhaps the default should be `null`?

pauloouriques replied:
It is confusing. I'll remove the comment because it's not helping.

evantahler commented on plugins/@grouparoo/mongo/src/initializers/plugin.ts:
```diff
@@ -0,0 +1,75 @@
+                "e.g. mongodb+srv://user:password@cluster0.123.mongodb.net/mydb?retryWrites=true&w=majority",
+            },
+            {
+              key: "database",
+              displayName: "Database name",
+              required: true,
```
I think I'm still confused if I should be providing database name in the URI or not.  This one says requried.

pauloouriques replied:
Yes, it's required.

evantahler replied:
Got it - so then shouldn't the connection string not include the database?

pauloouriques replied:
It's optional on the connection string.

evantahler commented on plugins/@grouparoo/mongo/src/initializers/plugin.ts:
```diff
@@ -0,0 +1,75 @@
+                "The MongoDB Connection String. Normally, this is connection string is provided by the Mongo DB service. " +
+                "In case you're using the standalone Mongo DB, this connection string will looks line this: mongodb://localhost:27017.",
+              placeholder:
+                "e.g. mongodb+srv://user:password@cluster0.123.mongodb.net/mydb?retryWrites=true&w=majority",
```
This differs from what you have in the template file.  Also, please remove the `e.g.`, because it will show literally this way in the UI

evantahler commented on plugins/@grouparoo/mongo/tsconfig.json:
```diff
@@ -0,0 +1,14 @@
+    "outDir": "./dist",
+    "allowJs": true,
+    "module": "commonjs",
+    "target": "es2018",
```
This is es2019 now

evantahler reviewed (changes requested):
Looking good! A few comments below:

pauloouriques committed:
chore: changes based on the PR comments.

grouparoo-bot committed:
Updating things automatically

evantahler commented:
A good guide on loading CSV data into Mongo https://kb.objectrocket.com/mongo-db/how-to-import-a-csv-into-mongodb-327

evantahler commented:
Looks good in the UI!

In a future story, we'll likely want to remove the need for a comma-separated list of object properties, it's a little odd. 
<img width="1585" alt="Screen Shot 2021-03-24 at 5 18 02 PM" src="https://user-images.githubusercontent.com/303226/112400222-1d1fc780-8cc5-11eb-815b-8320041ac7dd.png">
<img width="1585" alt="Screen Shot 2021-03-24 at 5 16 42 PM" src="https://user-images.githubusercontent.com/303226/112400226-201ab800-8cc5-11eb-9772-102363c9924e.png">

bleonard commented on .circleci/config.yml:
```diff
@@ -47,14 +47,14 @@ jobs:
-            - plugins/@grouparoo/sentry/node_modules
-            - plugins/@grouparoo/spec-helper/node_modules
-            - plugins/@grouparoo/zendesk/node_modules
+            - plugins/@grouparoo/mongo/node_modules
```
@evantahler  there would be less churn on these if, instead of `even` and `odd` we did something like alphabetize and then each has `length / 2`

Or maybe  set a max size for each (15?). and know the number of buckets (3). and then freak out when there are 4 buckets (>45)

bleonard replied:
Another PR of course

pauloouriques committed:
chore: rebase with main and bump version up.

evantahler committed:
update package versions

evantahler commented:
Rebasing off of main and bumping version numbers...
